# Phase 2: Data Cleaning

In [1]:
import pandas as pd

churn = pd.read_csv("../data/raw/customer_churn.csv")
customers = pd.read_csv("../data/raw/customers.csv")
engagement = pd.read_csv("../data/raw/customer_engagement_metrics.csv")
rfm = pd.read_csv("../data/raw/customer_rfm.csv")
orders = pd.read_csv("../data/raw/orders.csv")
billing = pd.read_csv("../data/raw/subscription_billing.csv")
tickets = pd.read_csv("../data/raw/support_tickets.csv")
campaign = pd.read_csv("../data/raw/campaign_responses.csv")
pd.set_option('display.max_columns',None)

In [2]:
datasets = {
    "customers": customers,
    "churn": churn,
    "engagement": engagement,
    "rfm": rfm,
    "orders": orders,
    "billing": billing,
    "tickets": tickets,
    "campaign": campaign
}

missing value treatment

| Dataset                     | Null Columns                 | Missing Values |
| --------------------------- | ---------------------------- | -------------- |
| customer_churn              | No missing values            | 0              |
| customers                   | country                      | 469            |
| customer_engagement_metrics | No missing values            | 0              |
| customer_rfm                | No missing values            | 0              |
| orders                      | shipping_country, promo_code | 573, 28616     |
| subscription_billing        | No missing values            | 0              |
| support_tickets             | No missing values            | 0              |
| campaign_responses          | No missing values            | 0              |


### country null value treatment

In [3]:
customers[customers['country'].isna()].head()

,customer_id,email,name,first_name,last_name,country,state,city,street_address,postal_code,signup_date,age,gender,occupation,marital_status,household_size,segment,lifetime_value,loyalty_points,preferred_channel,preferred_language,phone,account_status,referral_source,email_opt_in,sms_opt_in,last_login_date,created_at
108,C000109,tmcgee@example.com,Evelyn Galvan,Joseph,Robertson,NaN,IL,Savageton,0553 Rangel Forge,74727,2024-03-19,49,F,Corporate treasurer,single,4,Platinum,6192.75,8432,sms,fr,001-875-539-3733x056,active,organic,1,1,2026-02-13,2024-03-19
785,C000786,williamgriffith@example.com,Laurie Bailey,Robert,Montes,NaN,MS,Alvaradostad,58238 James Roads,45198,2023-01-06,65,F,Air traffic controller,married,4,Gold,10916.00,18665,sms,fr,715-738-7376,inactive,referral,0,1,2025-12-14,2023-01-06
1175,C001176,brucebradford@example.org,Janice Petty,Samantha,Leblanc,NaN,WY,Port Lindaland,98089 Claudia Locks,37994,2021-07-25,34,Other,Interior and spatial designer,married,3,Gold,11354.00,38308,sms,en,001-898-846-8896x37778,active,organic,0,1,2026-03-12,2021-07-25
1830,C001831,katherinemiles@example.com,Erika Daugherty,Brandon,Mayer,NaN,ID,South Paigeshire,81385 Waters Road Suite 047,6250,2021-12-01,20,M,Water quality scientist,divorced,6,Platinum,4390.00,16703,app,fr,(410)779-4190,suspended,partner,1,0,2025-11-27,2021-12-01
1858,C001859,aprilmoore@example.org,Erika Maxwell,Virginia,Evans,NaN,NV,North Jean,2955 John Square Suite 320,2372,2025-10-23,19,Other,Writer,widowed,2,Gold,24742.50,39805,sms,fr,302-696-7340,active,organic,1,0,2025-07-01,2025-10-23


In [4]:
customers['state'].nunique()

59

based on states we treat the null values

In [5]:
customers["country"] = customers["country"].fillna(
    customers.groupby("state")["country"]
             .transform(lambda x: x.mode()[0])
)

### orders null value treatment

In [6]:
orders[orders['shipping_country'].isna()].head()

,order_id,customer_id,order_date,channel,total_amount,subtotal,tax_amount,shipping_cost,discount_amount,status,payment_method,payment_last4,shipping_country,shipping_state,shipping_city,shipping_address,billing_city,carrier,tracking_number,promo_code,is_gift,fulfillment_center,delivery_date
106,O00000107,C051759,2026-06-23 10:25:00,mobile,619.27,544.96,37.16,19.82,12.39,returned,card,9738,NaN,IL,Savageton,0553 Rangel Forge,West Tracyside,USPS,TRKBB802474-88F,WELCOME,0,FC-EAST,2026-02-14
783,O00000784,C090819,2025-05-24 12:12:45,mobile,1091.60,960.61,65.50,34.93,21.83,cancelled,cod,3500,NaN,MS,Alvaradostad,58238 James Roads,Millerfurt,FedEx,TRK84B85236-9BC,WELCOME,0,FC-WEST,2026-03-28
1173,O00001174,C148339,2025-08-16 13:57:42,mobile,1135.40,999.15,68.12,36.33,22.71,returned,cod,4767,NaN,WY,Port Lindaland,98089 Claudia Locks,West Stevenside,FedEx,TRK22BBF468-CC1,VIP,0,FC-EAST,2026-02-24
1828,O00001829,C051438,2025-12-22 05:26:36,web,439.00,386.32,26.34,14.05,8.78,pending,wallet,7644,NaN,ID,South Paigeshire,81385 Waters Road Suite 047,Wilkinsonfort,FedEx,TRKB0A97817-43B,WELCOME,1,FC-WEST,2025-12-09
1856,O00001857,C047942,2025-12-23 22:22:56,mobile,2474.25,2177.34,148.46,79.18,49.49,cancelled,wallet,3133,NaN,NV,North Jean,2955 John Square Suite 320,Ashleyfurt,DHL,TRKDB463E79-587,WELCOME,1,FC-CENTRAL,2026-03-28


based on shipping state we treat null values for shipping country

In [7]:
orders["shipping_country"] = orders["shipping_country"].fillna(
    orders.groupby("shipping_state")["shipping_country"]
          .transform(lambda x: x.mode()[0])
)

In [8]:
orders[orders['promo_code'].isna()].head()

,order_id,customer_id,order_date,channel,total_amount,subtotal,tax_amount,shipping_cost,discount_amount,status,payment_method,payment_last4,shipping_country,shipping_state,shipping_city,shipping_address,billing_city,carrier,tracking_number,promo_code,is_gift,fulfillment_center,delivery_date
0,O00000001,C096901,2026-04-21 02:05:49,web,963.27,847.68,57.80,30.82,19.27,shipped,cod,7033,DK,GA,Danamouth,551 David Bypass Suite 955,Johnnyland,UPS,TRK36F5D547-166,NaN,0,FC-EAST,2026-02-25
5,O00000006,C091934,2025-06-07 07:41:24,marketplace,2499.00,2199.12,149.94,79.97,49.98,pending,card,3594,AG,GU,New Kenneth,3748 Michael Centers,Brandonshire,UPS,TRKD40A0CED-1D8,NaN,0,FC-EAST,2025-10-19
13,O00000014,C106782,2026-03-30 12:46:53,store,798.65,702.81,47.92,25.56,15.97,completed,card,5207,CI,MT,Pinedatown,62221 Stephanie Estates Suite 634,Harrisport,DHL,TRK78544253-D8A,NaN,1,FC-WEST,2026-04-17
14,O00000015,C114434,2026-02-22 20:43:15,web,390.57,343.71,23.43,12.50,7.81,completed,card,1682,SK,VT,Harrisport,4800 Davis Landing Apt. 451,Lindaborough,FedEx,TRK1F260BF9-D57,NaN,0,FC-CENTRAL,2026-04-22
16,O00000017,C119377,2024-12-07 06:49:29,marketplace,1613.40,1419.79,96.80,51.63,32.27,cancelled,cod,8268,BD,FL,Port Jason,5689 Brown Haven,East Sherri,UPS,TRKDAB327CB-AAF,NaN,0,FC-EAST,2026-03-18


In [9]:
orders["promo_code"] = orders["promo_code"].fillna(
    "No Promo"
)

In [10]:
for name, df in datasets.items():
        print(name)
        print(df.columns)
        print("-" * 30)

customers
Index(['customer_id', 'email', 'name', 'first_name', 'last_name', 'country',
       'state', 'city', 'street_address', 'postal_code', 'signup_date', 'age',
       'gender', 'occupation', 'marital_status', 'household_size', 'segment',
       'lifetime_value', 'loyalty_points', 'preferred_channel',
       'preferred_language', 'phone', 'account_status', 'referral_source',
       'email_opt_in', 'sms_opt_in', 'last_login_date', 'created_at'],
      dtype='object')
------------------------------
churn
Index(['customer_id', 'tenure_months', 'contract_type', 'monthly_spend',
       'days_since_last_order', 'support_tickets_30d', 'total_spend',
       'order_count', 'avg_order_value', 'payment_method', 'autopay',
       'num_products_owned', 'complaint_count', 'nps_score',
       'satisfaction_score', 'churned', 'churn_reason', 'predicted_churn_prob',
       'retention_offer_sent', 'offer_accepted', 'last_contact_date',
       'account_manager', 'updated_at'],
      dtype='object')


In [11]:

# Date Conversion


customers["signup_date"] = pd.to_datetime(customers["signup_date"])
customers["last_login_date"] = pd.to_datetime(customers["last_login_date"])
customers["created_at"] = pd.to_datetime(customers["created_at"])

engagement["last_login_date"] = pd.to_datetime(engagement["last_login_date"])
engagement["measured_date"] = pd.to_datetime(engagement["measured_date"])

orders["order_date"] = pd.to_datetime(orders["order_date"])
orders["delivery_date"] = pd.to_datetime(orders["delivery_date"])

billing["contract_end_date"] = pd.to_datetime(billing["contract_end_date"])
billing["created_at"] = pd.to_datetime(billing["created_at"])
billing["updated_at"] = pd.to_datetime(billing["updated_at"])

tickets["created_date"] = pd.to_datetime(tickets["created_date"])

churn["last_contact_date"] = pd.to_datetime(churn["last_contact_date"])
churn["updated_at"] = pd.to_datetime(churn["updated_at"])

In [12]:
for name, df in datasets.items():
    print(f"{name}: {df.duplicated().sum()}")

customers: 0
churn: 0
engagement: 0
rfm: 0
orders: 0
billing: 0
tickets: 0
campaign: 0


There are no completely identical rows in any dataset.

In [13]:
for name, df in datasets.items():
    if "customer_id" in df.columns:
        print(name)
        print("Unique:", df["customer_id"].nunique())
        print("Duplicates:", df["customer_id"].duplicated().sum())
        print("-"*30)

customers
Unique: 93551
Duplicates: 0
------------------------------
churn
Unique: 150000
Duplicates: 31175
------------------------------
engagement
Unique: 150000
Duplicates: 69237
------------------------------
rfm
Unique: 150000
Duplicates: 467852
------------------------------
orders
Unique: 80183
Duplicates: 34204
------------------------------
billing
Unique: 126941
Duplicates: 0
------------------------------
tickets
Unique: 72253
Duplicates: 26501
------------------------------
campaign
Unique: 146833
Duplicates: 431562
------------------------------


| Dataset    | Unique Customers | Duplicate IDs | Relationship | Action       |
| ---------- | ---------------: | ------------: | ------------ | ------------ |
| customers  |           93,551 |             0 | One-to-One   | Direct merge |
| churn      |          150,000 |        31,175 | One-to-Many  | Aggregate    |
| engagement |          150,000 |        69,237 | One-to-Many  | Aggregate    |
| rfm        |          150,000 |       467,852 | One-to-Many  | Aggregate    |
| orders     |           80,183 |        34,204 | One-to-Many  | Aggregate    |
| billing    |          126,941 |             0 | One-to-One   | Direct merge |
| tickets    |           72,253 |        26,501 | One-to-Many  | Aggregate    |
| campaign   |          146,833 |       431,562 | One-to-Many  | Aggregate    |


In [14]:
customers.to_csv(
    "../data/processed/customers_cleaned.csv",
    index=False
)

orders.to_csv(
    "../data/processed/orders_cleaned.csv",
    index=False
)